File: 070/eu070.go
// project euler (projecteuler.net) problem 70
// solution by Kevin Retzke (retzkek@gmail.com), May 2012
package main

import (
	"fmt"
	"math"
)

type Primes struct {
	Primes []int
	Last   int
}

// Init initializes a Primes struct with the first two primes.
func (p *Primes) Init() {
	p.Primes = []int{2, 3}
	p.Last = 3
}

// Next computes, returns, and appends the next prime number.
func (p *Primes) Next() int {
	next := 0
	i := p.Last + 2
	for next == 0 {
		sqrti := math.Sqrt(float64(i))
		isPrime := true
		for _, p := range p.Primes {
			if i%p == 0 {
				isPrime = false
				i += 2
				break
			}
			if float64(p) > sqrti {
				break
			}
		}
		if isPrime {
			next = i
		}
	}
	p.Primes = append(p.Primes, next)
	p.Last = next
	return next
}

// Eratosthenes populates p with all primes up to max, computed with
// the Sieve of Eratosthenes.
func (p *Primes) Eratosthenes(max int) {
    sieve := make([]bool, max)
    for i := 2; i*i < max; i++ {
        if !sieve[i] {
            for j := 2 * i; j < max; j += i {
                sieve[j] = true
            }
        }
    }
    p.Primes = []int{2}
    for i := 3; i < max; i++ {
        if !sieve[i] {
            p.Last = i
            p.Primes = append(p.Primes, p.Last)
        }
    }
}


func main() {
	p := new(Primes)
	p.Eratosthenes(1000000)
	minn := 999999 
	minnphin := float64(0)
	for n := 2; n <= 10000000; n += 2 {
		phi := float64(n)
		for _, m := range p.Primes {
			if n%m == 0 {
				phi *= (1 - 1/float64(m))
			}
		}
		if nphin := float64(n) / phi; nphin < minnphin {
			minn = n
			minnphin = nphin
			fmt.Println(minn, minnphin)
		}
	}
}

File: python/eu070.py
#!/usr/bin/python
# encoding: utf-8
"""
project euler (projecteuler.net) problem 69
solution by Kevin Retzke, April 2012

inefficient solution, max found so far:
300300 57600 5.21354166667
"""
import math

def gcd(n,m):
    """
    Calculate greatest common divisor of n and m using the 
    Euclidian algorithm.
    """
    if n > m:
        large = n
        small = m
    else:
        large = m
        small = n
    if n==0 or m==0:
        return large
    return gcd(large%small,small)

def phi(n):
    result = 0
    for k in xrange(1,n):
        if gcd(k,n) == 1:
            result += 1
    if result == 0:
        result = 1
    return result

def countDigits(n):
    """Returns list of len(10) with number of each digit in n."""
    digits = [0]*10
    while n > 0:
        digits[n%10] += 1
        n = n//10
    return digits

if __name__ == '__main__':
    maxij = 3163
    phis = {}
    minn = 0
    minnphin = 9999999.0
    for i in xrange(1,maxij+1):
        if i not in phis:
            phis[i] = phi(i)
        for j in xrange(i,maxij*maxij/i):
            if i==j==1:
                continue
            if j not in phis:
                phis[j] = phi(j)
            n = i*j
            gcdij = gcd(i,j)
            phin = phis[i]*phis[j]*gcdij/phis[gcdij]
            nphin = float(n)/phin
            if nphin < minnphin:
                if countDigits(n) == countDigits(phin):
                    minn = n
                    minnphin = nphin
                    print n, nphin
